### ISSUE TITLE
feat: more languages supported

### ISSUE THREAD

**Issue by elijah-potter** [elijah-potter/harper#79]
I'd like to expand the number of languages supported for comment checking. There are a number of additional considerations for structured comments (like JavaDoc), but some should be pretty straightforward.

Running list:
- Bash
- YAML
- HTML

Feel free to add suggestions to this list.

**Comment by elijah-potter** [comment]
Alright. At the time of writing, we've got JavaDoc, HTML, and Bash supported, as well as `gitcommit` buffers.

**Comment by mostafaqanbaryan** [comment]
If you could add `php` to the list as well, it would be great

**Comment by Vinzent03** [comment]
Adding `dart` would be great as well.

**Comment by kai-tub** [comment]
Hey, markdown support seems pretty advanced. :+1:
But are there plans to also support other markup languages, such as [djot](https://djot.net/), with the same level of functionality?
Or will only markdown be seen as a first-class citizen in the near future?

Either way, I am thrilled about the project and being able to use a solid spell-checker LSP :partying_face: :

PS: Are you planning on adding a sponsoring option?

**Comment by elijah-potter** [comment]
Coming back to this issue after a solid period of stagnation, to give an update.

#332 has added Haskell support, #354 is adding Vimscript, and #353 is adding `cmake` support. These will all be considered "rough" until someone has the time to go in and touch them up (which is a good first issue to tackle, if anyone is interested).

I plan on adding PHP and potentially PowerShell later today. 

> But are there plans to also support other markup languages, such as [djot](https://djot.net/), with the same level of functionality?

We've got an open PR for Typst support: #302. I do not _personally_ have time in the near future to add similar levels of polish to a new language as the Markdown parser. That said, we are entirely open to PRs on the topic, and I can answer any questions you may have. I'm not sure how complex djot is, but it may be as simple as importing the right Treesitter parser.

> Either way, I am thrilled about the project and being able to use a solid spell-checker LSP 🥳 

That's great to hear! I aim to please.

> PS: Are you planning on adding a sponsoring option?

[We might be a little past that.](https://techcrunch.com/2024/11/21/wordpress-com-owner-automattic-snaps-up-grammar-checker-harper)

**Comment by krupan** [comment]
Emacs org-mode support would be rad

**Comment by hippietrail** [comment]
I just had a peep at the supported languages and here are some I note to be missing:
- Assembly
- Elixir
- Kotlin
- Odin
- Zig

I've never used Elixir. Kotlin probably requires the least effort, possibly none. Zig and Odin from memory both also just have C-like comments. Assembly uses `;` to introduce comments, at least the common ones.

**Comment by jkprcl** [comment]
Glad to see PowerShell is on the running list!

**Comment by ficcdaf** [comment]
`mail` support isn't great right now. Since it uses the PlainEnglish parser, it also applies rules to the headers and metadata of the file which isn't ideal. I'm working on a [tree-sitter parser](https://github.com/ficcdaf/tree-sitter-mail) for the mail format atm, I'm wondering if it would be possible to only apply harper's rules to contents of the `(body)` node, or something like that?

**Comment by elijah-potter** [comment]
Yeah! Using the masker in `harper-tree-sitter` you can select by node label.

**Comment by ficcdaf** [comment]
sounds great. I will give that a shot when I'm able.

**Comment by funnes** [comment]
It would be nice to have support for [jujutsu](https://github.com/jj-vcs/jj) commits just like `git-commit`.

update: implemented in #2082

**Comment by ficcdaf** [comment]
Re: a discussion in the discord, we could look into supporting Quatro and RMarkdown. They're both extensions of markdown. RMarkdown has a very solid tree-sitter parser available. Quatro has one too but I haven't been able to verify if it still works. Using these parser and the Masker we should be able to support them. For the time being, they could be treated the same as regular Markdown files; after a brief test, this generally works fine at the cost of some false positives on certain syntax elements.

**Comment by derjasper** [comment]
I use VSCode to edit **LaTeX** documents, so having support for that would be nice!

**Comment by elijah-potter** [comment]
It looks like @grantlemons has started work on a Quatro parser. He previously attempted LaTeX, but it never made it in.

**Comment by grantlemons** [comment]
Yep, Quatro overall seems pretty easy to mask, but it extends Markdown in a few ways that make it non-trivial. (#749)

**Comment by ficcdaf** [comment]
> It looks like [@grantlemons](https://github.com/grantlemons) has started work on a Quatro parser. He previously attempted LaTeX, but it never made it in.

Masking Latex would definitely be non-trivial considering Latex's complexity and the turing-complete syntax. This is also why writing a tree-sitter parser that will correctly parse *all* possible Latex files is basically impossible.

I messed around with Latex parsing a while ago. Using the tree-sitter parser wouldn't be very consistent for masking because a lot of `(text)` nodes are actually the inputs to functions. It's pretty damn hard to consistently decide whether a given node or section of text should be checked or not. 

Ltex-ls is a project that does manage to do it, however. We could look at their implementation for some ideas; although I imagine adding LaTeX support would be no small feat.

**Comment by tymcauley** [comment]
I would love to have support for Scala and YAML. Thank you for such a great tool!

**Comment by anarion80** [comment]
Please add MDX support: https://mdxjs.com/docs/what-is-mdx/

**Comment by jacmoe** [comment]
If you are serious about Emacs support, then Org-mode is a must :)

Since I do all my writing in Org-mode, I can't use Harper, unfortunately. However, once support for org-mode is in, I will use it in a heartbeat!

**Comment by lopeztel** [comment]
Adding ReStructuredText (.rst) would be useful: https://docutils.sourceforge.io/docs/user/rst/quickstart.html

**Comment by 0xspryon** [comment]
Hi @elijah-potter , it'd be great if Harper had support for the [solidity programming language](https://soliditylang.org/).
I'd like to add it myself if that's possible. Would be great if you could point me to a few ressources/tips on how to go about it.
Thanks for creating this.
Cheers!

**Comment by mcecode** [comment]
> Would be great if you could point me to a few ressources/tips on how to go about it.

You can use #970 as a basis. To add comment checking for a new language, it's mostly about adding a tree-sitter grammar and some tests. Hope this helps.

**Comment by zjh-nuc-AIOT** [comment]
Can it support Verilog and SystemVerlog?

**Comment by hippietrail** [comment]
> Can it support Verilog and SystemVerlog?

I think the key ingredient is if there is a Tree-sitter grammar.

**Comment by zsmatrix62** [comment]
i would appreciate if html is supported as which is presented to users directly  🥲

**Comment by mcecode** [comment]
> i would appreciate if html is supported

It's already supported though, here's the list of [currently supported languages](https://writewithharper.com/docs/integrations/language-server#Supported-Languages). Do you mean you want HTML comment checking?

**Comment by Kiaazad** [comment]
I was about to ask how to enable this extension on .rpy files, but looks like it's not something that the user can do.
Perhaps instead of adding languages one by one, add a field that contains an editable list of file extensions to check. And do a general check just in strings and comments? Ignoring the code parts if those cause issue.

**Comment by mcecode** [comment]
> I was about to ask how to enable this extension on .rpy files, but looks like it's not something that the user can do. Perhaps instead of adding languages one by one, add a field that contains an editable list of file extensions to check. And do a general check just in strings and comments? Ignoring the code parts if those cause issue.

As much as this is a good idea, it isn't really possible. The reason we need to add support for languages one by one is because before `harper-ls` can lint anything, it first need to know how to parse the file it's given. Without knowing how to do that, `harper-ls` won't know what parts it needs to lint vs what parts it needs to ignore  (e.g., comments vs code).

**Comment by elijah-potter** [comment]
> I think the key ingredient is if there is a Tree-sitter grammar.

This is not entirely true. It's quite easy to add support _using_ a Tree-sitter grammar, but it can be done in normal Rust, too. We just need to be able to `Mask` out the parts of the document that _don't_ need to be checked.

**Comment by ficcdaf** [comment]
> Perhaps instead of adding languages one by one, add a field that contains an editable list of file extensions to check. And do a general check just in strings and comments? Ignoring the code parts if those cause issue.

One workaround could be to let the user configure Harper to treat certain file extensions (or maybe, as a fallback for any unsupported file types) as `PlainText`. This is obviously not ideal, because then Harper will lint _everything_, but for some use-cases, this may be better than nothing.

Alternatively, here's an idea: suppose we create a new masker which basically operates on regex. Users could provide regexes that define what should be `Mask`ed out, and Harper lints the rest. This way, users that are invested enough and feel like tinkering, could extend Harper to at least be usable for a new filetype without needing to wait on the upstream. What do you think @elijah-potter ?

**Comment by mcecode** [comment]
@pratyushmittal requested in #1373 to add support for the the Django template language (`htmldjango`). Upon a cursory look, I found [`tree-sitter-htmldjango`](https://github.com/interdependence/tree-sitter-htmldjango), which may help those who want to implement it. Though, when searching `crates.io`, there doesn't seem to be a create available for the grammar.

**Comment by elijah-potter** [comment]
Quick update: we just merged orgmode support in #1369.

> @pratyushmittal requested in https://github.com/Automattic/harper/discussions/1373 to add support for the the Django template language (htmldjango). Upon a cursory look, I found [tree-sitter-htmldjango](https://github.com/interdependence/tree-sitter-htmldjango), which may help those who want to implement it. Though, when searching crates.io, there doesn't seem to be a create available for the grammar.

The simplest solution might just be to request they publish it to `crates.io`.

**Comment by vazome** [comment]
Jupyter : `.ipynb`, `.ijlnb` and some other kernels are not currently supported, though it is basically `.{language}` + `.md`. Would be great to see the support one day

**Comment by jerielverissimo** [comment]
Hi, would be cool to support `Clojure` language too.

**Comment by iwahbe** [comment]
It would be great to have support for LISPs in general. I'm specifically looking for support for Common LISP and Emacs LISP, but since they are LISPs we should be able to get one grammar to more or less cover all of them.

**Comment by cfarm6** [comment]
It would be great to add `julia` too.

**Comment by nick-robo** [comment]
`R` too :)

**Comment by elijah-potter** [comment]
If anyone wants their first contribution to be language support, I've put together a small guide: https://elijahpotter.dev/articles/adding_a_programming_language_to_harper

**Comment by TealEgg** [comment]
> Support for the [asciidoc](https://asciidoc.org/) markup language would be really helpful.

Yeah I was just going to say that. ASCIIdoc imo is way better than Markdown. Wish it was more popular :)

**Comment by hippietrail** [comment]
How about `.rst`? I never heard of it until just now when I found some GitHub repos have `README.rst` instead of `README.md`: https://raw.githubusercontent.com/pybind/pybind11/refs/heads/master/README.rst

- https://github.com/orgs/community/discussions/86067
- https://en.wikipedia.org/wiki/ReStructuredText
- https://docutils.sourceforge.io/rst.html

**Comment by haotian9** [comment]
When do you add LaTex? Thanks.

**Comment by YourBrainEatsYou** [comment]
`vue sfc` would be nice

Thx

**Comment by hippietrail** [comment]
> When do you add LaTex? Thanks.

LaTeX has come up a number of times and if I got an accurate impression it's a notoriously difficult format to parse and not at all like most file formats.

**Comment by byryepez** [comment]
Hello! It would be very helpful if we could add other file extensions to existing file types. For example, Quarto documents (*.qmd) are essentially Markdown files (*.md), although they can include code chunks, similar to those found in a Jupyter notebook.

Thanks!

**Comment by elijah-potter** [comment]
@byryepez, if you'd like to open a PR, it should be as simple as adding a new mapping from `.qmd` to the Markdown parser and adding a few test cases.

**Comment by stan-ely** [comment]
@grantlemons Hi, I am just curious about your pull request https://github.com/Automattic/harper/pull/749 for quarto. Why did you create a masker in a separate crate, as opposed to the approach taken for org mode in https://github.com/Automattic/harper/pull/1369, by creating a new parser.

**Comment by grantlemons** [comment]
> @grantlemons Hi, I am just curious about your pull request https://github.com/Automattic/harper/pull/749 for quarto. Why did you create a masker in a separate crate, as opposed to the approach taken for org mode in https://github.com/Automattic/harper/pull/1369, by creating a new parser.

iirc it was mostly because it doesn't have to be part of harper-core.

I think decoupling the core from  language support is generally a good thing. In this case I even went farther and moved it to its own repository because it's pretty niche and doesn't need to be an 'officially' supported language.

**Comment by stan-ely** [comment]
@grantlemons Do we know which flavour of Markdown is currently supported? I noticed that headers aren’t supported at the moment.

In my previous experience with Hugo, I wrote content in .md files using GitHub Flavored Markdown. It’s quite similar to Quarto, apart from features like shortcodes.

If we could get the current Markdown implementation to support at least the basic syntax elements, like the header and code block, that would already be quite useful. We could then just call the markdown parser for quarto. What do you think?

**Comment by Crown0815** [comment]
Please add support for [asciidoc](https://asciidoc.org). For technical documentation it is far superior to markdown.

**Comment by ryan-ronnander** [comment]
> Please add support for [asciidoc](https://asciidoc.org). For technical documentation it is far superior to markdown.

Another vote for AsciiDoc. Checking grammar in AsciiDoc files was the reason I downloaded the plugin in the first place. The technical documentation I tend to work on is a blend of markdown and AsciiDoc.

**Comment by hippietrail** [comment]
Well there is a TreeSitter grammar for asciidoc: https://github.com/cathaysia/tree-sitter-asciidoc

**Comment by RoloEdits** [comment]
> It would be nice to have support for [jujutsu](https://github.com/jj-vcs/jj) commits just like `git-commit`.

+1 for this: here is the tree-sitter helix uses https://github.com/kareigu/tree-sitter-jjdescription

**Comment by calebeby** [comment]
I would love to see djot as well, as mentioned earlier. Its syntax is similar to markdown and there is a decent tree sitter parser available: https://github.com/treeman/tree-sitter-djot

**Comment by lloydjatkinson** [comment]
MDX please!

**Comment by eproxus** [comment]
Erlang support would be nice! There's a tree-sitter grammar at https://github.com/WhatsApp/tree-sitter-erlang

**Comment by matthewbegun** [comment]
> Hello! It would be very helpful if we could add other file extensions to existing file types. For example, Quarto documents (_.qmd) are essentially Markdown files (_.md), although they can include code chunks, similar to those found in a Jupyter notebook.
> 
> Thanks!

For anyone looking for Quarto support in VSCode, as a quick (but temporary) fix, you can just set the language to Markdown manually and then restart `harper-ls`. It won't recognise the extra Quarto features but it will work otherwise until you re-open the file or code.

**Comment by ariebh** [comment]
Support for R would be great. There is a tree-sitter grammar at [https://github.com/r-lib/tree-sitter-r](https://github.com/r-lib/tree-sitter-r)

**Comment by fmorroni** [comment]
+1 for zig

**Comment by Coadon** [comment]
Plaintex (latex) support would be awesome. It's essentially plaintext except for all the backslashed phrases and equations.

**Comment by Chris4942** [comment]
It would be great to get support for .proto files (protobufs)

https://protobuf.dev/

**Comment by samunemeth** [comment]
I think LaTeX would be amazing! In that case, this could replace `ltex-ls`.

**Comment by hippietrail** [comment]
> I think LaTeX would be amazing! In that case, this could replace `ltex-ls`.

- https://github.com/Automattic/harper/issues/56
- https://github.com/Automattic/harper/issues/2073

**Comment by LaBatata101** [comment]
+1 for Zig

**Comment by rishubn** [comment]
> > I think LaTeX would be amazing! In that case, this could replace `ltex-ls`.
> 
>     * [LaTeX Support #56](https://github.com/Automattic/harper/issues/56)
> 
>     * [Add LaTeX support #2073](https://github.com/Automattic/harper/pull/2073)

I think development is stalled due to license issues

**Comment by hippietrail** [comment]
> +1 for Zig

It was just added. Please test it.

**Comment by hippietrail** [comment]
This issue has become quite messy and hard to follow. I'm going to use it as a tracking issue and create child issues for individual languages which have been requested here but have not yet been implemented.

**Comment by Manuallly-Installed** [comment]
I would love it if suport for PureBasic was added.

**Comment by jtdowney** [comment]
Commenting to add that it would be great to see Gleam support, tree sitter is at https://github.com/gleam-lang/tree-sitter-gleam

**Comment by exiquio** [comment]
> I just had a peep at the supported languages and here are some I note to be missing:
> 
> * Assembly
> * Elixir
> * Kotlin
> * Odin
> * [Zig](https://github.com/Automattic/harper/issues/218)
> 
> I've never used Elixir. Kotlin probably requires the least effort, possibly none. Zig and Odin from memory both also just have C-like comments. Assembly uses `;` to introduce comments, at least the common ones.

Elixir would be nice to have.

**Comment by hippietrail** [comment]
> > I've never used Elixir. Kotlin probably requires the least effort, possibly none. Zig and Odin from memory both also just have C-like comments. Assembly uses `;` to introduce comments, at least the common ones.
> 
> Elixir would be nice to have.

There's actually an official Tree-sitter grammar for it too: https://github.com/elixir-lang/tree-sitter-elixir

**Comment by radioactive-exe** [comment]
I'm a bit surprised this hasn't been brought up yet, but if there could somehow be SQL support (or at least some of its dialects like T-SQL or Postgres), it would be wonderful!

**Comment by dccabanas** [comment]
Hey, all. It would be cool to have support for Julia :D

**Comment by grantlemons** [comment]
> Hey, all. It would be cool to have support for Julia :D

Tree sitter parser: https://github.com/tree-sitter/tree-sitter-julia
